File: multi-storage-transaction-sample/src/main/java/sample/Sample.java
package sample;

import com.scalar.db.api.DistributedTransaction;
import com.scalar.db.api.DistributedTransactionManager;
import com.scalar.db.api.Get;
import com.scalar.db.api.Put;
import com.scalar.db.api.Result;
import com.scalar.db.api.Scan;
import com.scalar.db.exception.transaction.TransactionException;
import com.scalar.db.io.Key;
import com.scalar.db.service.TransactionFactory;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

public class Sample implements AutoCloseable {

  private final DistributedTransactionManager manager;

  public Sample() throws IOException {
    // Create a transaction manager object
    TransactionFactory factory = TransactionFactory.create("database.properties");
    manager = factory.getTransactionManager();
  }

  public void loadInitialData() throws TransactionException {
    DistributedTransaction transaction = null;
    try {
      transaction = manager.start();
      loadCustomerIfNotExists(transaction, 1, "Yamada Taro", 10000, 0);
      loadCustomerIfNotExists(transaction, 2, "Yamada Hanako", 10000, 0);
      loadCustomerIfNotExists(transaction, 3, "Suzuki Ichiro", 10000, 0);
      loadItemIfNotExists(transaction, 1, "Apple", 1000);
      loadItemIfNotExists(transaction, 2, "Orange", 2000);
      loadItemIfNotExists(transaction, 3, "Grape", 2500);
      loadItemIfNotExists(transaction, 4, "Mango", 5000);
      loadItemIfNotExists(transaction, 5, "Melon", 3000);
      transaction.commit();
    } catch (TransactionException e) {
      if (transaction != null) {
        // If an error occurs, abort the transaction
        transaction.abort();
      }
      throw e;
    }
  }

  private void loadCustomerIfNotExists(
      DistributedTransaction transaction,
      int customerId,
      String name,
      int creditLimit,
      int creditTotal)
      throws TransactionException {
    Optional<Result> customer =
        transaction.get(
            Get.newBuilder()
                .namespace("customer")
                .table("customers")
                .partitionKey(Key.ofInt("customer_id", customerId))
                .build());
    if (!customer.isPresent()) {
      transaction.put(
          Put.newBuilder()
              .namespace("customer")
              .table("customers")
              .partitionKey(Key.ofInt("customer_id", customerId))
              .textValue("name", name)
              .intValue("credit_limit", creditLimit)
              .intValue("credit_total", creditTotal)
              .build());
    }
  }

  private void loadItemIfNotExists(
      DistributedTransaction transaction, int itemId, String name, int price)
      throws TransactionException {
    Optional<Result> item =
        transaction.get(
            Get.newBuilder()
                .namespace("order")
                .table("items")
                .partitionKey(Key.ofInt("item_id", itemId))
                .build());
    if (!item.isPresent()) {
      transaction.put(
          Put.newBuilder()
              .namespace("order")
              .table("items")
              .partitionKey(Key.ofInt("item_id", itemId))
              .textValue("name", name)
              .intValue("price", price)
              .build());
    }
  }

  public String getCustomerInfo(int customerId) throws TransactionException {
    DistributedTransaction transaction = null;
    try {
      // Start a transaction
      transaction = manager.start();

      // Retrieve the customer info for the specified customer ID from the customers table
      Optional<Result> customer =
          transaction.get(
              Get.newBuilder()
                  .namespace("customer")
                  .table("customers")
                  .partitionKey(Key.ofInt("customer_id", customerId))
                  .build());

      if (!customer.isPresent()) {
        // If the customer info the specified customer ID doesn't exist, throw an exception
        throw new RuntimeException("Customer not found");
      }

      // Commit the transaction (even when the transaction is read-only, we need to commit)
      transaction.commit();

      // Return the customer info as a JSON format
      return String.format(
          "{\"id\": %d, \"name\": \"%s\", \"credit_limit\": %d, \"credit_total\": %d}",
          customerId,
          customer.get().getText("name"),
          customer.get().getInt("credit_limit"),
          customer.get().getInt("credit_total"));
    } catch (Exception e) {
      if (transaction != null) {
        // If an error occurs, abort the transaction
        transaction.abort();
      }
      throw e;
    }
  }

  public String placeOrder(int customerId, int[] itemIds, int[] itemCounts)
      throws TransactionException {
    assert itemIds.length == itemCounts.length;

    DistributedTransaction transaction = null;
    try {
      String orderId = UUID.randomUUID().toString();

      // Start a transaction
      transaction = manager.start();

      // Put the order info into the orders table
      transaction.put(
          Put.newBuilder()
              .namespace("order")
              .table("orders")
              .partitionKey(Key.ofInt("customer_id", customerId))
              .clusteringKey(Key.ofBigInt("timestamp", System.currentTimeMillis()))
              .textValue("order_id", orderId)
              .build());

      int amount = 0;
      for (int i = 0; i < itemIds.length; i++) {
        int itemId = itemIds[i];
        int count = itemCounts[i];

        // Put the order statement into the statements table
        transaction.put(
            Put.newBuilder()
                .namespace("order")
                .table("statements")
                .partitionKey(Key.ofText("order_id", orderId))
                .clusteringKey(Key.ofInt("item_id", itemId))
                .intValue("count", count)
                .build());

        // Retrieve the item info from the items table
        Optional<Result> item =
            transaction.get(
                Get.newBuilder()
                    .namespace("order")
                    .table("items")
                    .partitionKey(Key.ofInt("item_id", itemId))
                    .build());

        if (!item.isPresent()) {
          throw new RuntimeException("Item not found");
        }

        // Calculate the total amount
        amount += item.get().getInt("price") * count;
      }

      // Check if the credit total exceeds the credit limit after payment
      Optional<Result> customer =
          transaction.get(
              Get.newBuilder()
                  .namespace("customer")
                  .table("customers")
                  .partitionKey(Key.ofInt("customer_id", customerId))
                  .build());
      if (!customer.isPresent()) {
        throw new RuntimeException("Customer not found");
      }
      int creditLimit = customer.get().getInt("credit_limit");
      int creditTotal = customer.get().getInt("credit_total");
      if (creditTotal + amount > creditLimit) {
        throw new RuntimeException("Credit limit exceeded");
      }

      // Update credit_total for the customer
      transaction.put(
          Put.newBuilder()
              .namespace("customer")
              .table("customers")
              .partitionKey(Key.ofInt("customer_id", customerId))
              .intValue("credit_total", creditTotal + amount)
              .build());

      // Commit the transaction
      transaction.commit();

      // Return the order id
      return String.format("{\"order_id\": \"%s\"}", orderId);
    } catch (Exception e) {
      if (transaction != null) {
        // If an error occurs, abort the transaction
        transaction.abort();
      }
      throw e;
    }
  }

  private String getOrderJson(DistributedTransaction transaction, String orderId)
      throws TransactionException {
    // Retrieve the order info for the order ID from the orders table
    Optional<Result> order =
        transaction.get(
            Get.newBuilder()
                .namespace("order")
                .table("orders")
                .indexKey(Key.ofText("order_id", orderId))
                .build());

    if (!order.isPresent()) {
      throw new RuntimeException("Order not found");
    }

    int customerId = order.get().getInt("customer_id");

    // Retrieve the customer info for the specified customer ID from the customers table
    Optional<Result> customer =
        transaction.get(
            Get.newBuilder()
                .namespace("customer")
                .table("customers")
                .partitionKey(Key.ofInt("customer_id", customerId))
                .build());
    assert customer.isPresent();

    // Retrieve the order statements for the order ID from the statements table
    List<Result> statements =
        transaction.scan(
            Scan.newBuilder()
                .namespace("order")
                .table("statements")
                .partitionKey(Key.ofText("order_id", orderId))
                .build());

    // Make the statements JSONs
    List<String> statementJsons = new ArrayList<>();
    int total = 0;
    for (Result statement : statements) {
      int itemId = statement.getInt("item_id");

      // Retrieve the item data from the items table
      Optional<Result> item =
          transaction.get(
              Get.newBuilder()
                  .namespace("order")
                  .table("items")
                  .partitionKey(Key.ofInt("item_id", itemId))
                  .build());

      if (!item.isPresent()) {
        throw new RuntimeException("Item not found");
      }

      int price = item.get().getInt("price");
      int count = statement.getInt("count");

      statementJsons.add(
          String.format(
              "{\"item_id\": %d,\"item_name\": \"%s\",\"price\": %d,\"count\": %d,\"total\": %d}",
              itemId, item.get().getText("name"), price, count, price * count));

      total += price * count;
    }

    // Return the order info as a JSON format
    return String.format(
        "{\"order_id\": \"%s\",\"timestamp\": %d,\"customer_id\": %d,\"customer_name\": \"%s\",\"statement\": [%s],\"total\": %d}",
        orderId,
        order.get().getBigInt("timestamp"),
        customerId,
        customer.get().getText("name"),
        String.join(",", statementJsons),
        total);
  }

  public String getOrderByOrderId(String orderId) throws TransactionException {
    DistributedTransaction transaction = null;
    try {
      // Start a transaction
      transaction = manager.start();

      // Get an order JSON for the specified order ID
      String orderJson = getOrderJson(transaction, orderId);

      // Commit the transaction (even when the transaction is read-only, we need to commit)
      transaction.commit();

      // Return the order info as a JSON format
      return String.format("{\"order\": %s}", orderJson);
    } catch (Exception e) {
      if (transaction != null) {
        // If an error occurs, abort the transaction
        transaction.abort();
      }
      throw e;
    }
  }

  public String getOrdersByCustomerId(int customerId) throws TransactionException {
    DistributedTransaction transaction = null;
    try {
      // Start a transaction
      transaction = manager.start();

      // Retrieve the order info for the customer ID from the orders table
      List<Result> orders =
          transaction.scan(
              Scan.newBuilder()
                  .namespace("order")
                  .table("orders")
                  .partitionKey(Key.ofInt("customer_id", customerId))
                  .build());

      // Make order JSONs for the orders of the customer
      List<String> orderJsons = new ArrayList<>();
      for (Result order : orders) {
        orderJsons.add(getOrderJson(transaction, order.getText("order_id")));
      }

      // Commit the transaction (even when the transaction is read-only, we need to commit)
      transaction.commit();

      // Return the order info as a JSON format
      return String.format("{\"order\": [%s]}", String.join(",", orderJsons));
    } catch (Exception e) {
      if (transaction != null) {
        // If an error occurs, abort the transaction
        transaction.abort();
      }
      throw e;
    }
  }

  public void repayment(int customerId, int amount) throws TransactionException {
    DistributedTransaction transaction = null;
    try {
      // Start a transaction
      transaction = manager.start();

      // Retrieve the customer info for the specified customer ID from the customers table
      Optional<Result> customer =
          transaction.get(
              Get.newBuilder()
                  .namespace("customer")
                  .table("customers")
                  .partitionKey(Key.ofInt("customer_id", customerId))
                  .build());
      if (!customer.isPresent()) {
        throw new RuntimeException("Customer not found");
      }

      int updatedCreditTotal = customer.get().getInt("credit_total") - amount;

      // Check if over repayment or not
      if (updatedCreditTotal < 0) {
        throw new RuntimeException("Over repayment");
      }

      // Reduce credit_total for the customer
      transaction.put(
          Put.newBuilder()
              .namespace("customer")
              .table("customers")
              .partitionKey(Key.ofInt("customer_id", customerId))
              .intValue("credit_total", updatedCreditTotal)
              .build());

      // Commit the transaction
      transaction.commit();
    } catch (Exception e) {
      if (transaction != null) {
        // If an error occurs, abort the transaction
        transaction.abort();
      }
      throw e;
    }
  }

  @Override
  public void close() {
    manager.close();
  }
}

File: scalardb-dotnet-samples/scalardb-cluster-sql-sample/Sample.cs
using System.Diagnostics;
using System.Text.Json.Nodes;
using Microsoft.Extensions.Logging;
using ScalarDB.Client;
using ScalarDB.Client.Builders.Sql;
using ScalarDB.Client.Exceptions;

namespace ScalarDbClusterSqlSample;

public class Sample: IDisposable
{
    private readonly ISqlTransactionManager _manager;

    public Sample()
    {
        var loggerFactory = LoggerFactory.Create(builder =>
        {
            builder.SetMinimumLevel(LogLevel.Warning);
            builder.AddSimpleConsole(options => { options.TimestampFormat = "HH:mm:ss "; });
        });
        
        var factory = TransactionFactory.Create("scalardb-options.json", loggerFactory);
        _manager = factory.GetSqlTransactionManager();
    }

    public async Task CreateTables()
    {
        await _manager.ExecuteAsync("CREATE COORDINATOR TABLES IF NOT EXIST");
        await _manager.ExecuteAsync("CREATE NAMESPACE IF NOT EXISTS sample");

        await _manager.ExecuteAsync("CREATE TABLE IF NOT EXISTS sample.customers (" +
                                    "customer_id INT PRIMARY KEY, " +
                                    "name TEXT, " +
                                    "credit_limit INT, " +
                                    "credit_total INT" +
                                    ")");
        await _manager.ExecuteAsync("CREATE TABLE IF NOT EXISTS sample.orders (" +
                                    "customer_id INT, " +
                                    "\"timestamp\" BIGINT, " +
                                    "order_id TEXT, " +
                                    "PRIMARY KEY (customer_id, \"timestamp\")" +
                                    ")");
        await _manager.ExecuteAsync("CREATE TABLE IF NOT EXISTS sample.statements (" +
                                    "order_id TEXT, " +
                                    "item_id INT, " +
                                    "count INT, " +
                                    "PRIMARY KEY (order_id, item_id)" +
                                    ")");
        await _manager.ExecuteAsync("CREATE TABLE IF NOT EXISTS sample.items (" +
                                    "item_id INT PRIMARY KEY, " +
                                    "name TEXT, " +
                                    "price INT" +
                                    ")");
        
        await _manager.ExecuteAsync("CREATE INDEX IF NOT EXISTS ON sample.orders (order_id)");
    }

    public async Task LoadInitialData()
    {
        var transaction = await _manager.BeginAsync();
        try
        {
            await LoadCustomerIfNotExists(1, "Yamada Taro", 10000, 0);
            await LoadCustomerIfNotExists(2, "Yamada Hanako", 10000, 0);
            await LoadCustomerIfNotExists(3, "Suzuki Ichiro", 10000, 0);
            await LoadItemIfNotExists(1, "Apple", 1000);
            await LoadItemIfNotExists(2, "Orange", 2000);
            await LoadItemIfNotExists(3, "Grape", 2500);
            await LoadItemIfNotExists(4, "Mango", 5000);
            await LoadItemIfNotExists(5, "Melon", 3000);
            
            await transaction.CommitAsync();
        }
        catch (TransactionException)
        {
            // If an error occurs, rollback the transaction
            await transaction.RollbackAsync();
            throw;
        }
        
        return;

        async Task LoadCustomerIfNotExists(int customerId, string name,
                                           int creditLimit, int creditTotal)
        {
            var resultSet =
                await transaction.ExecuteAsync(
                    new SqlStatementBuilder()
                        .SetSql("SELECT * FROM sample.customers WHERE customer_id = ?")
                        .AddParam(customerId)
                        .Build()
                );

            if (resultSet.Records.Count > 0)
                return;

            await transaction.ExecuteAsync(
                new SqlStatementBuilder()
                    .SetSql("INSERT INTO sample.customers (customer_id, name, credit_limit, credit_total) VALUES (?, ?, ?, ?)")
                    .AddParam(customerId)
                    .AddParam(name)
                    .AddParam(creditLimit)
                    .AddParam(creditTotal)
                    .Build()
            );
        }

        async Task LoadItemIfNotExists(int itemId, string name, int price)
        {
            var resultSet =
                await transaction.ExecuteAsync(
                    new SqlStatementBuilder()
                        .SetSql("SELECT * FROM sample.items WHERE item_id = ?")
                        .AddParam(itemId)
                        .Build()
                );
            
            if (resultSet.Records.Count > 0)
                return;

            await transaction.ExecuteAsync(
                new SqlStatementBuilder()
                    .SetSql("INSERT INTO sample.items (item_id, name, price) VALUES (?, ?, ?)")
                    .AddParam(itemId)
                    .AddParam(name)
                    .AddParam(price)
                    .Build()
            );
        }
    }

    public async Task<JsonObject> GetCustomerInfo(int customerId)
    {
        var transaction = await _manager.BeginAsync();
        try
        {
            // Retrieve the customer info for the specified customer ID from the customers table
            var resultSet =
                await transaction.ExecuteAsync(
                    new SqlStatementBuilder()
                        .SetSql("SELECT * FROM sample.customers WHERE customer_id = ?")
                        .AddParam(customerId)
                        .Build()
                );

            var customer = resultSet.Records.FirstOrDefault();
            if (customer is null)
            {
                // If the customer info the specified customer ID doesn't exist, throw an exception
                throw new Exception($"Customer not found (id: {customerId})");
            }

            // Commit the transaction (even when the transaction is read-only, we need to commit)
            await transaction.CommitAsync();

            return new JsonObject
                   {
                       { "id", customerId },
                       { "name", customer.GetValue<string>("name") },
                       { "credit_limit", customer.GetValue<int>("credit_limit") },
                       { "credit_total", customer.GetValue<int>("credit_total") }
                   };
        }
        catch (TransactionException)
        {
            // If an error occurs, rollback the transaction
            await transaction.RollbackAsync();
            throw;
        }
    }

    public async Task<JsonObject> PlaceOrder(int customerId,
                                             IReadOnlyDictionary<int, int> itemCounts)
    {
        var transaction = await _manager.BeginAsync();
        try
        {
            var orderId = Guid.NewGuid().ToString();

            // Insert the order info into the orders table
            await transaction.ExecuteAsync(
                new SqlStatementBuilder()
                    .SetSql("INSERT INTO sample.orders (customer_id, order_id, \"timestamp\") VALUES (?, ?, ?)")
                    .AddParam(customerId)
                    .AddParam(orderId)
                    .AddParam(DateTimeOffset.UtcNow.ToUnixTimeMilliseconds())
                    .Build()
            );

            var amount = 0;
            foreach (var (itemId, count) in itemCounts)
            {
                // Insert the order statement into the statements table
                await transaction.ExecuteAsync(
                    new SqlStatementBuilder()
                        .SetSql("INSERT INTO sample.statements (order_id, item_id, count) VALUES (?, ?, ?)")
                        .AddParam(orderId)
                        .AddParam(itemId)
                        .AddParam(count)
                        .Build()
                );

                // Retrieve the item info from the items table
                var result =
                    await transaction.ExecuteAsync(
                        new SqlStatementBuilder()
                            .SetSql("SELECT * FROM sample.items WHERE item_id = ?")
                            .AddParam(itemId)
                            .Build()
                    );
 
                var item = result.Records.FirstOrDefault();
                if (item is null)
                    throw new Exception($"Item not found (id: {itemId})");

                // Calculate the total amount
                amount += item.GetValue<int>("price") * count;
            }

            // Check if the credit total exceeds the credit limit after payment
            var resultSet =
                await transaction.ExecuteAsync(
                    new SqlStatementBuilder()
                        .SetSql("SELECT * FROM sample.customers WHERE customer_id = ?")
                        .AddParam(customerId)
                        .Build()
                );

            var customer = resultSet.Records.FirstOrDefault();
            if (customer is null)
                throw new Exception($"Customer not found (id: {customerId})");

            var creditLimit = customer.GetValue<int>("credit_limit");
            var creditTotal = customer.GetValue<int>("credit_total");
            var newCreditTotal = creditTotal + amount;
            if (newCreditTotal > creditLimit)
                throw new Exception($"Credit limit exceeded ({newCreditTotal} > {creditLimit})");

            // Update credit_total for the customer
            await transaction.ExecuteAsync(
                new SqlStatementBuilder()
                    .SetSql("UPDATE sample.customers SET credit_total = ? WHERE customer_id = ?")
                    .AddParam(newCreditTotal)
                    .AddParam(customerId)
                    .Build()
            );

            // Commit the transaction
            await transaction.CommitAsync();

            return new JsonObject
                   {
                       { "order_id", orderId }
                   };
        }
        catch (TransactionException)
        {
            // If an error occurs, rollback the transaction
            await transaction.RollbackAsync();
            throw;
        }
    }

    public async Task<JsonObject> GetOrderByOrderId(string orderId)
    {
        var transaction = await _manager.BeginAsync();
        try
        {
            // Get an order JSON for the specified order ID
            var orderJson = await getOrder(transaction, orderId);

            // Commit the transaction (even when the transaction is read-only, we need to commit)
            await transaction.CommitAsync();

            return new JsonObject
                   {
                       { "order", orderJson }
                   };
        }
        catch (TransactionException)
        {
            // If an error occurs, rollback the transaction
            await transaction.RollbackAsync();
            throw;
        }
    }

    public async Task<JsonObject> GetOrdersByCustomerId(int customerId)
    {
        var transaction = await _manager.BeginAsync();
        try
        {
            // Retrieve the order info for the customer ID from the orders table
            var resultSet =
                await transaction.ExecuteAsync(
                    new SqlStatementBuilder()
                        .SetSql("SELECT * FROM sample.orders WHERE customer_id = ?")
                        .AddParam(customerId)
                        .Build()
                );

            // Make order JSONs for the orders of the customer
            var orderJsons = new JsonArray();
            foreach (var order in resultSet.Records)
                orderJsons.Add(await getOrder(transaction, order.GetValue<string>("order_id")));

            // Commit the transaction (even when the transaction is read-only, we need to commit)
            await transaction.CommitAsync();

            return new JsonObject
                   {
                       { "orders", orderJsons }
                   };
        }
        catch (TransactionException)
        {
            // If an error occurs, rollback the transaction
            await transaction.RollbackAsync();
            throw;
        }
    }

    public async Task Repayment(int customerId, int amount)
    {
        var transaction = await _manager.BeginAsync();
        try
        {
            // Retrieve the customer info for the specified customer ID from the customers table
            var resultSet =
                await transaction.ExecuteAsync(
                    new SqlStatementBuilder()
                        .SetSql("SELECT * FROM sample.customers WHERE customer_id = ?")
                        .AddParam(customerId)
                        .Build()
                );

            var customer = resultSet.Records.FirstOrDefault();
            if (customer is null)
                throw new Exception($"Customer not found (id: {customerId})");

            var updatedCreditTotal = customer.GetValue<int>("credit_total") - amount;

            // Check if over repayment or not
            if (updatedCreditTotal < 0)
                throw new Exception($"Over-repayment ({updatedCreditTotal})");

            // Reduce credit_total for the customer
            await transaction.ExecuteAsync(
                new SqlStatementBuilder()
                    .SetSql("UPDATE sample.customers SET credit_total = ? WHERE customer_id = ?")
                    .AddParam(updatedCreditTotal)
                    .AddParam(customerId)
                    .Build()
            );

            // Commit the transaction
            await transaction.CommitAsync();
        }
        catch (TransactionException)
        {
            // If an error occurs, rollback the transaction
            await transaction.RollbackAsync();
            throw;
        }
    }

    private static async Task<JsonObject> getOrder(ISqlTransaction transaction, string orderId)
    {
        // Retrieve the order info for the order ID from the orders table
        var resultSet =
            await transaction.ExecuteAsync(
                new SqlStatementBuilder()
                    .SetSql("SELECT * FROM sample.orders WHERE order_id = ?")
                    .AddParam(orderId)
                    .Build()
            );
  
        var order = resultSet.Records.FirstOrDefault();
        if (order is null)
            throw new Exception($"Order not found (id: {orderId})");

        var customerId = order.GetValue<int>("customer_id");

        // Retrieve the customer info for the specified customer ID from the customers table
        resultSet =
            await transaction.ExecuteAsync(
                new SqlStatementBuilder()
                    .SetSql("SELECT * FROM sample.customers WHERE customer_id = ?")
                    .AddParam(customerId)
                    .Build()
            );

        var customer = resultSet.Records.FirstOrDefault();
        Debug.Assert(customer is not null);

        // Retrieve the order statements for the order ID from the statements table
        resultSet =
            await transaction.ExecuteAsync(
                new SqlStatementBuilder()
                    .SetSql("SELECT * FROM sample.statements WHERE order_id = ?")
                    .AddParam(orderId)
                    .Build()
            );
        var statements = resultSet.Records;

        // Make the statements JSONs
        var total = 0;
        var statementJsons = new JsonArray();

        foreach (var statement in statements)
        {
            var itemId = statement.GetValue<int>("item_id");

            // Retrieve the item data from the items table
            resultSet =
                await transaction.ExecuteAsync(
                    new SqlStatementBuilder()
                        .SetSql("SELECT * FROM sample.items WHERE item_id = ?")
                        .AddParam(itemId)
                        .Build()
                );

            var item = resultSet.Records.FirstOrDefault();
            if (item is null)
                throw new Exception($"Item not found (id: {itemId})");

            var price = item.GetValue<int>("price");
            var count = statement.GetValue<int>("count");
            var totalForStatement = price * count;

            statementJsons.Add(
                new JsonObject
                {
                    { "item_id", itemId },
                    { "item_name", item.GetValue<string>("name") },
                    { "price", price },
                    { "count", count },
                    { "total", totalForStatement }
                }
            );

            total += totalForStatement;
        }
        
        return new JsonObject
               {
                   { "order_id", orderId },
                   { "timestamp", order.GetValue<long>("timestamp") },
                   { "customer_id", customerId },
                   { "customer_name", customer.GetValue<string>("name") },
                   { "statements", statementJsons },
                   { "total", total }
               };
    }

    public void Dispose()
        => _manager.Dispose();
}
